File: src/main/java/com/refinedmods/refinedstorage/inventory/item/FilterIconItemHandler.java
package com.refinedmods.refinedstorage.inventory.item;

import com.refinedmods.refinedstorage.item.FilterItem;
import net.minecraft.item.ItemStack;
import net.minecraftforge.items.ItemStackHandler;

public class FilterIconItemHandler extends ItemStackHandler {
    private final ItemStack stack;

    public FilterIconItemHandler(ItemStack stack) {
        super(1);

        this.stack = stack;

        setStackInSlot(0, FilterItem.getIcon(stack));
    }

    @Override
    protected void onContentsChanged(int slot) {
        super.onContentsChanged(slot);

        FilterItem.setIcon(stack, getStackInSlot(0));
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/inventory/item/FilterIconItemHandler.kt
package com.refinedmods.refinedstorage.inventory.item

import com.refinedmods.refinedstorage.item.FilterItem
import net.minecraft.item.ItemStack
import net.minecraftforge.items.ItemStackHandler

class FilterIconItemHandler(private val stack: ItemStack) : ItemStackHandler(1) {
    protected fun onContentsChanged(slot: Int) {
        super.onContentsChanged(slot)
        FilterItem.setIcon(stack, getStackInSlot(0))
    }

    init {
        setStackInSlot(0, FilterItem.getIcon(stack))
    }
}